packages/lodestar-beacon-state-transition/src/fast/util/epochContext.ts

```diff
@@ -43,19 +43,19 @@ export class EpochContext {
   // Warning: may contain indices that do not yet exist in the current state, but do in a later processed state.
   public index2pubkey: PublicKey[];
   public proposers: number[];
-  public previousShuffling?: IEpochShuffling;
-  public currentShuffling?: IEpochShuffling;
-  public nextShuffling?: IEpochShuffling;
+  public previousShuffling: IEpochShuffling;
+  public currentShuffling: IEpochShuffling;
+  public nextShuffling: IEpochShuffling;
   public config: IBeaconConfig;
+  public state: BeaconState;
 
-  constructor(config: IBeaconConfig) {
+  constructor(config: IBeaconConfig, state: BeaconState) {
     this.config = config;
+    this.state = state;
```

Comment:
Is there any problem with keeping the state as an internal property? Otherwise, is there a better way to do the `copy` method below without keeping an internal reference to the state?

Reply:
I don't think we want to keep the state as an internal property. We could make the constructor look like:
`constructor({config, state, epochCtx}: {config: IBeaconConfig; state?: BeaconState; epochCtx?: EpochContext})`, and switch behavior to either copy or loadState, depending on if `state` or `epochCtx` is provided.

Tbh this feels strange, though I guess this is buying us that we won't be able to misuse the EpochContext.

What if we did something like

```typescript
interface EpochContext {
  shuffling: EpochShuffling;
  ...
}

interface UninitializedEpochContext {
  shuffling?: EpochShuffling;
  // no copy method
}

class EpochContext {
  // don't use directly
  constructor(config) {}
  create(config): UninitializedEpochContext {}
  loadState(state): EpochContext {}
  copy(): EpochContext {}
}
```